packages/response-ops/rule_form/src/hooks/use_rule_form_steps.tsx

```diff
@@ -0,0 +1,300 @@
+  );
+
+  const stepOrder: RuleFormStepId[] = useMemo(
+    () => STEP_ORDER.filter((stepId) => steps[stepId]),
+    [steps]
+  );
+
+  return { steps, stepOrder };
+};
+
+interface RuleFormVerticalSteps {
+  steps: EuiStepsProps['steps'];
+}
+
+export const useRuleFormSteps: () => RuleFormVerticalSteps = () => {
+  // Track steps that the user has interacted with and then focused away from
+  const [touchedSteps, setTouchedSteps] = useState<Record<RuleFormStepId, boolean>>(
+    STEP_ORDER.reduce(
+      (result, stepId) => ({ ...result, [stepId]: false }),
+      {} as Record<RuleFormStepId, boolean>
+    )
+  );
+
+  const reportOnBlur = useCallback(
+    (stepId: RuleFormStepId, element: React.JSX.Element) => (
+      <div
+        data-test-subj={`ruleFormStep-${stepId}-reportOnBlur`}
+        onBlur={() =>
+          !touchedSteps[stepId] &&
+          setTouchedSteps({
+            ...touchedSteps,
+            [stepId]: true,
+          })
+        }
+      >
+        {element}
+      </div>
+    ),
+    [touchedSteps]
+  );
+
+  const { steps, stepOrder } = useCommonRuleFormSteps({ touchedSteps });
+
+  const mappedSteps = useMemo(() => {
+    return stepOrder
+      .map((stepId) => {
+        const step = steps[stepId];
+        return step
+          ? {
+              ...step,
+              children: reportOnBlur(stepId, step.children),
+            }
+          : null;
+      })
+      .filter(Boolean) as EuiStepsProps['steps'];
+  }, [steps, stepOrder, reportOnBlur]);
+
+  return { steps: mappedSteps };
+};
+
+interface RuleFormHorizontalSteps {
+  steps: EuiStepsHorizontalProps['steps'];
+  currentStepComponent: React.ReactNode;
+  goToNextStep: () => void;
+  goToPreviousStep: () => void;
+  hasNextStep: boolean;
+  hasPreviousStep: boolean;
+}
+export const useRuleFormHorizontalSteps: () => RuleFormHorizontalSteps = () => {
+  const [currentStep, setCurrentStep] = useState<RuleFormStepId>(STEP_ORDER[0]);
+  const [touchedSteps, setTouchedSteps] = useState<Record<RuleFormStepId, boolean>>(
+    STEP_ORDER.reduce(
+      (result, stepId) => ({ ...result, [stepId]: false }),
+      {} as Record<RuleFormStepId, boolean>
+    )
+  );
+
+  const { steps, stepOrder } = useCommonRuleFormSteps({
+    touchedSteps,
+    currentStep,
+  });
+
+  const hasNextStep = useMemo(() => {
+    if (currentStep) {
+      const currentIndex = stepOrder.indexOf(currentStep);
+      return currentIndex < stepOrder.length - 1;
+    }
+    return false;
+  }, [stepOrder, currentStep]);
+  const hasPreviousStep = useMemo(() => {
+    if (currentStep) {
+      const currentIndex = stepOrder.indexOf(currentStep);
+      return currentIndex > 0;
+    }
+    return false;
+  }, [stepOrder, currentStep]);
+  const goToNextStep = useCallback(() => {
+    if (currentStep && hasNextStep) {
+      const currentIndex = stepOrder.indexOf(currentStep);
+      const nextStep = stepOrder[currentIndex + 1];
```

Comment:
Should we take into account the situation where `currentIndex + 1 > stepOrder.length // out of array boundaries`?

Reply:
That's handled by including `hasNextStep` in the `if` block

Reply:
Good point, I missed that. Thanks!